drivers/chibios/spi_master.c

```diff
@@ -115,21 +115,47 @@ bool spi_start(pin_t slavePin, bool lsbFirst, uint8_t mode, uint16_t divisor) {
     return true;
 }
 
-spi_status_t spi_write(uint8_t data) { return spi_transmit(&data, 1); }
+spi_status_t spi_write(uint8_t data) {
+    uint8_t rxData;
+    spiExchange(&SPI_DRIVER, 1, &data, &rxData);
+
+    return rxData;
+}
 
 spi_status_t spi_read(void) {
     uint8_t data = 0;
-    spi_receive(&data, 1);
+    spiReceive(&SPI_DRIVER, 1, &data);
+
     return data;
 }
 
 spi_status_t spi_transmit(const uint8_t *data, uint16_t length) {
-    spiSend(&SPI_DRIVER, length, data);
+    spi_status_t status;
+
+    for (uint16_t i = 0; i < length; i++) {
+        status = spi_write(data[i]);
+
+        if (status < 0) {
+            return status;
+        }
+    }
+
```

Comment:
This completely destroys the use of DMA for bulk transfers... before and after this PR gets applied (with SPI running at 40MHz):

[![Slow SPI Transfers](https://img.youtube.com/vi/dsqcR5OiVSw/0.jpg)](https://youtu.be/dsqcR5OiVSw)

SPI has no concept of "failures" once the master has acquired the SPI subsystem, it sends data out at that point - there's no collisions or whatnot to deal with. If the slave responds, great... but the SPI peripheral on the MCU will read the data on MISO as it's clocked regardless of if the slave is actually responding or not.

I'd suggest rolling back the bulk transfer functions, just keeping the single-byte transfer change you've made above. If we need to read back synchronous responses, then we should really add an `spi_exchange` function for it.

Reply:
Good point, done.